File: pms.py
import argparse
import matplotlib.pyplot as plt
import numpy as np
np.set_printoptions(suppress=True, precision=4)
import utils
import ls

if __name__ == "__main__":
    
    cli = argparse.ArgumentParser()
    cli.add_argument(
        "--kernel_threshold", action="store",
        default= 1e3, help="set the kernel threshold value")
    cli.add_argument(
        "--damping", action="store",
        default= 1, help="set the damping value"
    )
    cli.add_argument(
        "--start_num_iterations", action="store",
        default= 6, help="set the number of iterations to do in the partial least square"
    )
    cli.add_argument(
        "--num_iterations", action="store",
        default= 19, help="set the number of iterations to do in the least square"
    )
    
    args = cli.parse_args()
    
    print(f"[Reading trajectory datas...]\n")
    traj_meas,traj_gt= utils.read_trajectory('./data/trajectoy.dat');
    pose_dim= 3 
    num_poses= traj_meas.shape[1]
    print(f"[Trajectory datas correctly read]\n") 
    
    print(f"[Converting robot poses and trajectory gt to homogeneous matrices...]\n")
    XR_guess= utils.convert_robot_poses(num_poses, traj_meas)
    XR_true= utils.convert_robot_gt(num_poses,traj_gt)
    print(f"[Converted correctly]\n")
    
    print(f"[Reading landmarks positions datas...]\n")
    XL_true= utils.readLandmarksGT('./data/world.dat'); 
    num_landmarks= XL_true.shape[1]
    landmark_dim= XL_true.shape[0]
    print(f"[Positions correctly read]\n")
    
    print(f"[Reading camera datas...]\n")
    K, cam_pose, z_near, z_far, image_cols, image_rows= utils.read_camera("./data/camera.dat")
    print(f"[Camera datas correctly read]\n")
    
    Zr= utils.import_poses(num_poses, XR_guess)    
    projection_associations,Zp= utils.import_projections(num_poses, num_landmarks)
    id_landmarks = np.unique(projection_associations[1,:])
    num_landmarks = len(id_landmarks)
    print(f"[Strating landmark initialization...]\n")
    XL_guess,Zp,projection_associations,id_landmarks = utils.init_landmarks(XR_guess,Zp,projection_associations,id_landmarks, num_poses, num_landmarks, K, cam_pose)
    num_landmarks = len(id_landmarks)
    XL_guess=np.squeeze(XL_guess, axis=2)
    print(f"[Landmarks coorectly initialized]\n")
    
    print(f"[Starting preliminary landmarks optimization...]\n")
    XR_guess1,XL_guess1, chi_stats_p, num_inliers_p, chi_stats_r, num_inliers_r, H, b= ls.ls_warmup(XR_guess,XL_guess,Zp,projection_associations,Zr,args.start_num_iterations,args.damping,args.kernel_threshold, num_poses, pose_dim, num_landmarks, landmark_dim, K, cam_pose, z_near, image_rows, image_cols)
    print(f"[Landmarks optimization done]\n")
    
    print(f"[Starting Least Square...]\n")
    XR,XL, chi_stats_p, num_inliers_p, chi_stats_r, num_inliers_r, H, b= ls.total_ls(XR_guess1,XL_guess1,Zp,projection_associations,Zr,args.num_iterations,args.damping,args.kernel_threshold, num_poses, pose_dim, num_landmarks, landmark_dim, K, cam_pose, z_near, image_rows, image_cols)
    print(f"[Least Square done]\n")
    
    utils.show_results(XR_true, XR_guess, XR, XL_true, XL_guess, XL, chi_stats_r,num_inliers_r,chi_stats_p,num_inliers_p)
    utils.evaluate_poses(XR, XR_true)

File: PMS.m
close all
clear
clc


# Read odometry and ground truth
function [traj_meas,traj_gt] = readTrajectory(file_path)
  fid = fopen(file_path, 'r');
  data = textscan(fid, "%d %f %f %f %f %f %f");
  traj_meas = cell2mat(data(1,2:4))';
  traj_gt = cell2mat(data(1,5:7))';
  fclose(fid);
endfunction


# Read camera params
function [cam_mat, cam_trans, z_near, z_far, width, height] = readCamera(file_path)
  fid = fopen(file_path, 'r');
  data = textscan(fid, "%f %f %f", 3, "HeaderLines",1, "CollectOutput",1);
  cam_mat = cell2mat(data);
  data = textscan(fid, "%f %f %f %f", 4, "HeaderLines",1, "CollectOutput",1);
  cam_trans = cell2mat(data);
  data = textscan(fid, "%*s %f");
  data = cell2mat(data);
  z_near = data(1);
  z_far = data(2);
  width = data(3);
  height = data(4);
  fclose(fid);
endfunction

# Plot odometry and ground truth
function plotOdometryAndGT(traj_meas, traj_gt)
  subplot(2,1,1)
  plt=plot(traj_meas(:,1), traj_meas(:,2))
  title("Measurements")
  subplot(2,1,2)
  plt=plot(traj_gt(:,1), traj_gt(:,2))
  title("Ground Truth")
  waitfor(plt)
endfunction

#Read landmark true positions
function lan_gt = readLandmarksGT(file_path)
  fid = fopen(file_path, 'r');
  data = textscan(fid, "%*d %f %f %f");
  lan_gt = cell2mat(data)'';
  fclose(fid);
endfunction

function [id_landmarks, measurements] = readMeasurements(i)
  i = num2str(i-1,'%05.f'); % i-1 since measurements start from 0
  fid = fopen(strcat("./data/meas-",i,".dat"), 'r');
  data = textscan(fid, "%*s %*d %f %f %f", "HeaderLines",3);
  id_landmarks = cell2mat(data(1,1))';
  measurements = cell2mat(data(1,2:3))';
  fclose(fid);
endfunction

function A=v2t(v)
  c=cos(v(3));
  s=sin(v(3));
	A=[c, -s, v(1);
	   s,  c, v(2);
	   0,  0,   1];
end

# Read the trajectory (measurements & ground truth)
[traj_meas,traj_gt] = readTrajectory("./data/trajectoy.dat");
global num_poses = size(traj_meas,2)
global pose_dim = size(traj_meas,1)
#num_poses

#traj_meas(:,1)
#v2t(traj_meas(:,1))

XR_guess = zeros(3,3,num_poses);
for i = 1:num_poses
  XR_guess(:,:,i) = v2t(traj_meas(:,i));
end
#XR_guess()

# Read camera parameters
#[cam_mat, cam_trans, z_near, z_far, width, height] = readCamera("./data/camera.dat")
#lan_gt = readLandmarksGT("./data/world.dat")

#plotOdometryAndGT(traj_meas, traj_gt);

%%%%%%%%%%%%%%% POSE MEASUREMENTS %%%%%%%%%%%%%%%
Zr=zeros(3,3,num_poses-1);

for measurement_num=1:num_poses-1
  Xi=XR_guess(:,:,measurement_num);
  Xj=XR_guess(:,:,measurement_num+1);
  Zr(:,:,measurement_num)=inv(Xi)*Xj;
end

Zr(:,:,1)